File: Week_03/G20200343030400/LeetCode_127_400.cpp
//127
//https://leetcode-cn.com/problems/word-ladder

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        int wordSize = beginWord.length();
        unordered_map<string, vector<string>> comDict;
        unordered_map<string, bool> visited;
        for (string word : wordList) {
            for (int i = 0; i < wordSize; i++) {
                string comWord = word.substr(0, i) + '*' + word.substr(i + 1, wordSize);
                comDict[comWord].push_back(word);
            }
            visited[word] = false;
        }
        queue<pair<string, int>> qWords;
        qWords.push(make_pair(beginWord, 1));
        visited[beginWord] = true;
        while (!qWords.empty()) {
            pair<string, int> node = qWords.front();
            qWords.pop();
            string word = node.first;
            int level = node.second;
            for (int i = 0; i < wordSize; i++) {
                string comWord = word.substr(0, i) + '*' + word.substr(i + 1, wordSize);
                for (string adWord : comDict[comWord]) {
                    if (adWord == endWord)
                        return level + 1;
                    if (!visited[adWord]) {
                        visited[adWord] = true;
                        qWords.push(make_pair(adWord, level + 1));
                    }
                }
            }
        }
        return 0;
    }
};

File: Week_06/G20200343030400/LeetCode_127_400.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList: return 0

        wordList = set(wordList)
        word_len = len(beginWord)
        beginSet = {beginWord}
        endSet = {endWord}
        dist = 1
        
        while beginSet:
            dist += 1
            nextSet = set()

            for word in beginSet:
                for i in range(word_len):
                    for c in string.ascii_lowercase:
                        if c != word[i]:
                            new_word = word[:i] + c + word[i + 1:]
                            if new_word in endSet:
                                return dist
                            if new_word in wordList:
                                nextSet.add(new_word)
                                wordList.remove(new_word)
            
            beginSet = nextSet
            if len(beginSet) > len(endSet):
                beginSet, endSet = endSet, beginSet

        return 0
